File: Examples/Rust/src/examples/usb_connection.rs
use super::connection;
use crate::helpers;
use ximu3::connection_info::*;
use ximu3::port_scanner::*;

pub fn run() {
    if helpers::yes_or_no("Search for connections?") {
        let devices = PortScanner::scan_filter(PortType::Usb);

        let Some(device) = devices.first() else {
            println!("No USB connections available");
            return;
        };

        println!("Found {device}");

        let connection_info = &device.connection_info;

        connection::run(connection_info);
    } else {
        let connection_info = &ConnectionInfo::UsbConnectionInfo(UsbConnectionInfo {
            port_name: "COM1".to_owned(),
        }); // replace with actual connection info

        connection::run(connection_info);
    }
}

File: Examples/C/examples/usb_connection.c
#include "../helpers.h"
#include "connection.h"
#include <stdio.h>
#include "Ximu3.h"

void usb_connection() {
    if (yes_or_no("Search for connections?")) {
        const XIMU3_Devices devices = XIMU3_port_scanner_scan_filter(XIMU3_PortTypeUsb);

        if (devices.length == 0) {
            printf("No USB connections available\n");
            return;
        }
        printf("Found %s %s\n", devices.array[0].device_name, devices.array[0].serial_number);

        const XIMU3_UsbConnectionInfo connection_info = devices.array[0].usb_connection_info;

        XIMU3_devices_free(devices);

        XIMU3_Connection *const connection = XIMU3_connection_new_usb(connection_info);

        run(connection, XIMU3_usb_connection_info_to_string(connection_info));
    } else {
        const XIMU3_UsbConnectionInfo connection_info = (XIMU3_UsbConnectionInfo){
            .port_name = "COM1",
        }; // replace with actual connection info

        XIMU3_Connection *const connection = XIMU3_connection_new_usb(connection_info);

        run(connection, XIMU3_usb_connection_info_to_string(connection_info));
    }
}
